twine/repository.py

```diff
@@ -91,12 +98,21 @@ def close(self) -> None:
 
     @staticmethod
     def _convert_data_to_list_of_tuples(data: Dict[str, Any]) -> List[Tuple[str, Any]]:
+        # This does what ``warehouse.forklift.parse_form_metadata()`` does, in reverse.
         data_to_send = []
         for key, value in data.items():
-            if key in KEYWORDS_TO_NOT_FLATTEN or not isinstance(value, (list, tuple)):
-                data_to_send.append((key, value))
-            else:
+            key = KEYWORDS_TO_RENAME.get(key, key)
+            if isinstance(value, (list, tuple)) and key not in KEYWORDS_TO_NOT_FLATTEN:
                 data_to_send.extend((key, item) for item in value)
+            elif key == "project_urls":
+                assert isinstance(value, dict), "for mypy"
+                for name, url in value:
+                    data_to_send.append((key, f"{name}, {url}"))
+            elif key == "keywords":
+                assert isinstance(value, list), "for mypy"
+                data_to_send.append((key, ", ".join(value)))
+            else:
+                data_to_send.append((key, value))
```

Comment:
I'm concerned about how this code replaces code that was previously largely agnostic to the specific keys. Now it contains two special cases. Can we refactor this so it behaved more or less how it did before (i.e. either flatten or don't)... or maybe let each of the data.items provide a rendering function? I'd like to avoid entangling concerns, and this logic has a lot of entanglement. For example, `keywords` is reached because it's not in KEYWORDS_TO_NOT_FLATTEN, but `project_urls` is reached because it's not a list or tuple. I'd like to get back to an implementation that's less complicated.

Reply:
The code already contained special case in the handling of a few keys. It was just hiding the special keys into a module global variable. Now there are more special cases, but not that many more. The names of the metadata fields in `pyproject.toml`, in `METADATA` and in the submission form are not consistent. Thus there needs to be a rename somewhere. Then there are fields that are represented differently in the form, and there are two cases in which `packaging.metadata.parse_email()` decomposes a structured field into its components for validation and the code here needs to put them back into textual representation.

Changing the code to have all the values in `PackageMetadata` to be object with a serialization method seems much more complex that what this function does. The logic to implement would still need to be the same, but in that implementation it would be diluted into boilerplate code. But maybe I'm misunderstanding what you are suggesting.

I can change the order of the conditionals to have `keywords` handled first, so that the logic is a bit simpler.

Reply:
I've restructured the code. It is not easier to read and it is evident that only two special cases are added to the one existing.